### PR TITLE
Checks execution abilities

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

This PR, implements abilities on both frontend and backend for checks executions.
The user can execute checks on both hosts and clusters only if the correct ability is present in the user abilities.

- `all:host_checks_execution` enable host check execution
- `all:cluster_checks_execution` enable cluster check execution

Recording 2024-06-28 at 10 47 50

The backend returns forbidden when the user has not the correct ability.
Frontend buttons are disabled when the user has not the correct ability.

## How was this tested?

Automated tests

### CODE REVIEW

File: priv/repo/migrations/20240626142058_add_check_execution_permissions.exs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  use Ecto.Migration
+
+  def up do
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'cluster_checks_execution', 'Permits all operations on the Cluster checks executions', NOW(), NOW())"
+
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'hosts_checks_execution', 'Permits all operations on the Host checks executions', NOW(), NOW())"
+  end
+
+  def down do
+    execute "DELETE FROM abilities WHERE abilities.name = 'all' and abilities.resource = 'cluster_checks_execution'"
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:11]: It's the same, it's an old habit when I was writing a lot of plain sql queries kek
When we have one table it's the same

File: test/trento_web/controllers/v1/cluster_controller_test.exs [hunk 2]
```diff
@@ -47,6 +47,37 @@ defmodule TrentoWeb.V1.ClusterControllerTest do
   end
 
   describe "request check executions" do
+    test "should perform the request when the user has all:cluster_checks_execution ability", %{
+      conn: conn
+    } do
+      %{id: cluster_id} = insert(:cluster)
+
+      %{id: user_id} = insert(:user)
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply test/trento_web/controllers/v1/cluster_controller_test.exs:55]: I was thinking about that, but I want an "empty" "new user", the user created in the test helper has the global ability, so I have to modify the old abilities, delete them and then add the new, I think it's less obscure to have a new plain user and add the only necessary things other then mutating the "global" test user

File: priv/repo/migrations/20240626142058_add_check_execution_permissions.exs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+defmodule Trento.Repo.Migrations.AddCheckExecutionPermissions do
+  use Ecto.Migration
+
+  def up do
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'cluster_checks_execution', 'Permits all operations on the Cluster checks executions', NOW(), NOW())"
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:5]: done

File: priv/repo/migrations/20240626142058_add_check_execution_permissions.exs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'hosts_checks_execution', 'Permits all operations on the Host checks executions', NOW(), NOW())"
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:7]: nice catch!

Done

File: lib/trento/hosts/projections/host_read_model.ex [hunk 5]
```diff
@@ -13,6 +13,8 @@ defmodule Trento.Hosts.Projections.HostReadModel do
   alias Trento.Hosts.Projections.SlesSubscriptionReadModel
   alias Trento.Tags.Tag
 
+  defdelegate authorize(action, user, params), to: Trento.Hosts.Policy
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/hosts/projections/host_read_model.ex:16]: done

File: test/trento/hosts/policy_test.exs [hunk 6]
```diff
@@ -21,8 +21,26 @@ defmodule Trento.Hosts.PolicyTest do
   test "should allow unguarded actions" do
     user = %User{abilities: []}
 
-    Enum.each([:list, :delete, :request_checks_execution], fn action ->
+    Enum.each([:list, :delete], fn action ->
       assert Policy.authorize(action, user, HostReadModel)
     end)
   end
+
+  test "should allow request_checks_execution action if the user has all:hosts_checks_execution ability" do
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply test/trento/hosts/policy_test.exs:29]: done

File: test/trento/hosts/policy_test.exs [hunk 7]
```diff
@@ -21,8 +21,26 @@ defmodule Trento.Hosts.PolicyTest do
...
+    user = %User{abilities: [%Ability{name: "all", resource: "hosts_checks_execution"}]}
+
+    assert true == Policy.authorize(:request_checks_execution, user, HostReadModel)
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply test/trento/hosts/policy_test.exs:32]: I always prefer to be explicit, but it's true we can omit them, less code

File: test/trento_web/controllers/v1/host_controller_test.exs [hunk 8]
```diff
@@ -13,6 +13,45 @@ defmodule TrentoWeb.V1.HostControllerTest do
   setup :setup_api_spec_v1
   setup :setup_user
 
+  describe "forbidden routes" do
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply test/trento_web/controllers/v1/host_controller_test.exs:16]: added

File: assets/js/pages/ChecksSelection/ChecksSelectionHeader.jsx [hunk 9]
```diff
@@ -53,20 +54,25 @@ function ChecksSelectionHeader({
               visible={isAbleToStartExecution}
               wrap={false}
             >
-              <Button
-                type="primary"
-                className="mx-1"
-                onClick={onStartExecution}
-                disabled={!isAbleToStartExecution}
+              <DisabledGuard
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply assets/js/pages/ChecksSelection/ChecksSelectionHeader.jsx:57]: done

File: assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx [hunk 10]
```diff
@@ -30,6 +30,7 @@ describe('ChecksSelectionHeader component', () => {
         selection={selection}
         userAbilities={[{ name: 'all', resource: 'all' }]}
         checkSelectionPermittedFor={['all:all']}
+        checkExecutionPermittedFor={['all:all']}
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx:33]: done

File: test/trento_web/controllers/v1/host_controller_test.exs (hunk 8 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply test/trento_web/controllers/v1/host_controller_test.exs:16]: Could we still put this describe at the end of the file and as `forbidden response`?
Just to be aligned with the other controller test files.
We have the same for the `cluster_controller`

File: assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx [hunk 11]
```diff
@@ -146,6 +149,40 @@ describe('ChecksSelectionHeader component', () => {
     }
   );
 
+  it('should forbid starting a check execution', async () => {
```
Comment (arbulu89) [inline assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx:152]: This test looks wrong.
Here we should have the button disabled and the tooltip, as we do in the next test.
So we shouldn't be passing `all:all` in the userAbilities

**Review by arbulu89 — APPROVED** [review]
Thank you!
All good.
I think there is something wrong in one FE test

File: assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx (hunk 11 above)
Comment (CDimonaco) [reply assets/js/pages/ChecksSelection/ChecksSelectionHeader.test.jsx:152]: thanks for spotting, this morning I'm a little bit sleepy 😴

File: priv/repo/migrations/20240626142058_add_check_execution_permissions.exs [hunk 12]
```diff
@@ -0,0 +1,15 @@
+defmodule Trento.Repo.Migrations.AddCheckExecutionPermissions do
+  use Ecto.Migration
+
+  def up do
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'cluster_checks_execution', 'Permits all operations on cluster checks executions', NOW(), NOW())"
+
+    execute "INSERT INTO abilities(id, name, resource, label, inserted_at, updated_at) VALUES (DEFAULT, 'all', 'host_checks_execution', 'Permits all operations on host checks executions', NOW(), NOW())"
```
Comment (EMaksy) [inline priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:7]: NOW(), NOW() is for inserted_at and updated_at?

Comment (CDimonaco) [reply priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:7]: yes

**Review by EMaksy — APPROVED** [review]
LGTM, just left one question out of uncertainty :+1:

File: priv/repo/migrations/20240626142058_add_check_execution_permissions.exs (hunk 12 above)
Comment (arbulu89) [reply priv/repo/migrations/20240626142058_add_check_execution_permissions.exs:7]: yes

**Pull request merged**